### PR TITLE
[Distributed Strategy] Support independent RoPE type for HCA/CSA layers

### PR DESCRIPTION
**Pull request by JiabinYang**
### PR Category
Distributed Strategy

### PR Types
New features

### Description
Allow HCA (`csa_compress_ratios == 128`) and CSA (`2 <= ratio < 128`) layers to independently choose the RoPE variant (plain RoPE vs YaRN), instead of the previous fixed behavior where every compressed layer used YaRN.

**New `TransformerConfig` fields (both default `None` -> fully backward compatible):**
- `hca_rope_type` / `csa_rope_type`: `"rope"` | `"yarn"` | `None`

The positional-embedding width is deliberately *not* part of this PR: every layer keeps using the global `qk_pos_emb_head_dim`.

**Changes:**
- `transformer_config.py`: declare the two fields, add `transform_rules` mappings, and validate the values in `__post_init__` (`{"rope", "yarn"}` or `None`).
- `dsv4_hybrid_attention.py`: replace the hardcoded `use_compressed_yarn = compress_ratio > 1` with a per-attention-type resolution that keeps the historical default (yarn for compressed layers, plain rope for window/MQA) when the field is unset.

**Tests:** new `tests/single_card_tests/transformer/test_hca_csa_independent_rope.py` (17 tests) covering config defaults/validation, per-type resolution, and — most importantly — **bit-wise equivalence with the legacy code path**: a `mock.patch` shim reproduces exactly what the pre-PR code built when its local `use_compressed_yarn` flag was forced to `False`, and the test asserts the rotary module (`inv_freq`), the forward output and the input gradient are all identical for HCA and CSA layers (GPU forward + backward, same weights via `set_state_dict`, same input). A vacuity guard checks that the default YaRN path really does produce a different output, so the equality above cannot pass trivially.

**Results:** `test_hca_csa_independent_rope.py` **17 passed**; together with the regressions `test_dsv4_hybrid_attention` + `test_csa_train_indexer_only` **118 passed**. `pre-commit` (ruff check / ruff format / typos / copyright) clean on the touched files.

**Backward compatibility:** with both fields unset (`None`), behavior is byte-identical to before — compressed layers use YaRN with the global `qk_pos_emb_head_dim`.

### CODE REVIEW

File: src/paddlefleet/transformer/transformer_config.py [hunk 1]
```diff
@@ -1676,6 +1704,26 @@ def __post_init__(self):
...
+                val = getattr(self, name, None)
+                if val is not None and val not in valid_rope_types:
+                    raise ValueError(
+                        f"{name}={val!r} is invalid. Must be one of "
+                        "{'rope', 'yarn'} or None (keep default)."
+                    )
+            for name in ("hca_qk_pos_emb_head_dim", "csa_qk_pos_emb_head_dim"):
+                val = getattr(self, name, None)
+                if val is not None and (
+                    not isinstance(val, int) or isinstance(val, bool) or val < 0
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py]: P1 这里仅校验了非负整数，但该值随后会直接作为 RoPE 维度使用。

例如 `csa_qk_pos_emb_head_dim=1` 时，`RotaryEmbedding` 会生成 2 个频率分量，而 `dsv4_hybrid_attention.py` 的前向只切出 1 维；设置大于 `v_head_dim`（CSA 还必须不超过 `dsa_index_head_dim`）则 `nope_dim` 为负，主压缩器和索引器的 `_apply_rope` 路径都会在前向发生 shape 错误。当前“>= 0”契约会接受这些值并延迟到运行时失败。

请在配置校验中要求 0 或偶数，并限制 HCA 不超过 `v_head_dim`、CSA 不超过 `min(v_head_dim, dsa_index_head_dim)`；同时补一条实际前向回归，而不只断言构造后的属性。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [独立 RoPE 维度讨论](https://github.com/PaddlePaddle/PaddleFleet/pull/1673#discussion_r3736446663) | P1 | ✅ |
| 2 | [PR 标题/描述](https://github.com/PaddlePaddle/PaddleFleet/pull/1673) | P3 | ✅ |

本轮结论：原 P1 已随功能范围收缩而消除，P3 元数据已与当前实现同步。

- P1：当前提交已移除按注意力类型设置 RoPE 维度的能力，所有层继续使用全局 `qk_pos_emb_head_dim`，原维度边界问题不再存在。
- P3：标题和描述现仅表述 HCA/CSA 独立选择 RoPE 类型，与当前实现一致。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/transformer_config.py (hunk 1 above)
Comment (JiabinYang) [reply src/paddlefleet/transformer/transformer_config.py]: Thanks, fixed in 00d8fa41.

The `__post_init__` validation for `hca_qk_pos_emb_head_dim` / `csa_qk_pos_emb_head_dim` is no longer just a `>= 0` check. It now enforces the constraints these dims must satisfy before they reach the forward RoPE math:

- **Even (or 0):** RoPE rotates channel pairs (`paddle.chunk(x, 2, axis=-1)` in `_rotate_half`, and `RotaryEmbedding` builds `inv_freq` over `arange(0, dim, 2)`), so an odd width splits unevenly. The dim must now be `0` (NoPE) or even.
- **Upper bound:** the value is used as the RoPE width where `nope_dim = head_dim - qk_pos_emb_head_dim`. HCA feeds only the main compressor (`head_dim == v_head_dim`), so `hca_qk_pos_emb_head_dim` must be `<= v_head_dim`. CSA additionally threads the same dim into the Lightning Indexer compressor (`head_dim == dsa_index_head_dim`), so `csa_qk_pos_emb_head_dim` must be `<= min(v_head_dim, dsa_index_head_dim)`. This turns the deferred forward shape error (negative `nope_dim`) into an eager, explicit config error.

I also added an actual forward regression (`TestPerTypeDimForwardRegression`) that runs the DSv4 hybrid attention forward+backward on GPU for HCA and CSA layers with a custom even dim, the NoPE (dim=0) case, and the boundary `dim == head_dim` case, instead of only asserting constructed attributes. Plus config-level tests for the odd/oversized rejections. All 30 tests pass and pre-commit is clean.

Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_config.py]: 已在当前提交复核：配置校验现已拒绝奇数维度及超出 HCA `v_head_dim` 或 CSA `min(v_head_dim, dsa_index_head_dim)` 的值，并补充了 GPU 前向/反向回归。该 P1 已解决。

**Review by risemeup1111 — APPROVED** [review]

**Comment by JiabinYang** [comment]
Thanks @risemeup1111 for the review and approval.

Note (edited after force-push to `86a47f50`): the scope of this PR has been **narrowed**. It now adds only **two** optional, default-`None` fields to `src/paddlefleet/transformer/transformer_config.py`:

- `hca_rope_type`
- `csa_rope_type`

The two per-attention-type RoPE **dimension** fields mentioned in the original version of this comment (`hca_qk_pos_emb_head_dim` / `csa_qk_pos_emb_head_dim`) and their even/upper-bound validation have been dropped entirely, so the earlier P1 discussion no longer applies to the current diff. All layers keep using the global `qk_pos_emb_head_dim`.

The only remaining gate is `Check approval`, which requires an approval from one of @sneaxiy / @From00 / @ForFishes for the `transformer_config.py` change (+24/-0). Could one of you help take a look? Existing configs are unaffected: with both fields unset (`None`), behavior is identical to before. Thanks!

**Comment by risemeup1111** [comment]
已在当前 head 完成复核并确认这项变更可接受，我的批准已公开保留。`Check approval` 要求仓库指定审批人之一的批准，这部分需要由 @sneaxiy、@From00 或 @ForFishes 处理，我无法替代该门禁。

**Review by risemeup1111 — APPROVED** [review]

**Comment by JiabinYang** [comment]
Status update after force-push to `86a47f50`:

- **Scope narrowed.** This PR now only adds `hca_rope_type` / `csa_rope_type` (both optional, default `None`) so HCA / CSA layers can independently pick plain RoPE vs YaRN. The per-attention-type RoPE *dimension* fields from the previous revision were removed; every layer keeps using the global `qk_pos_emb_head_dim`. Diff is 3 files, +490/-8, of which 440 lines are the new test file.
- **Equivalence is now pinned by a test.** `test_hca_csa_independent_rope.py` (17 tests) asserts that `hca_rope_type="rope"` / `csa_rope_type="rope"` reproduces the pre-PR code path with its local `use_compressed_yarn` flag forced to `False`, bit for bit: same `inv_freq`, and identical forward output + input gradient on GPU (same weights via `set_state_dict`, same input). A vacuity guard checks the default YaRN path really does differ, so the equality cannot pass trivially.
- @risemeup1111 has re-approved the new head.

Remaining gate is `Check approval`: it needs one of @sneaxiy / @From00 / @ForFishes to approve the `src/paddlefleet/transformer/transformer_config.py` change. Could one of you take a look when you have a moment? Thanks!

**Comment by risemeup1111** [comment]
已确认当前 `86a47f50` 与收缩后的功能范围一致，我的批准仍公开有效，复核后没有新增 finding。`Check approval` 仍需由仓库指定审批人 @sneaxiy、@From00 或 @ForFishes 之一完成，我无法替代该门禁。

**Comment by JiabinYang** [comment]
Thanks for the re-confirmation @risemeup1111 — much appreciated. Understood on the `Check approval` gate; we'll wait for one of the designated approvers. Nothing further needed from your side.

**Comment by risemeup1111** [comment]
收到，当前批准保持不变。

**Review by sneaxiy — APPROVED** [review]
